python/ray/experimental/async_plasma.py

```diff
@@ -36,19 +33,30 @@ def close(self):
                 fut.cancel()
 
     def _complete_future(self, futures: List[asyncio.Future], ray_object_id):
+        logger.debug(
+            "Completing plasma futures for object id {}".format(ray_object_id))
+
         obj = ray.get(ray_object_id)
         for fut in futures:
-            try:
-                fut.set_result(obj)
-            except asyncio.InvalidStateError:
-                # Avoid issues where process_notifications
-                # and check_ready both get executed
-                pass
-        try:
-            self._waiting_dict.pop(ray_object_id)
-        except KeyError:
-            # Same situation as above
-            pass
+            loop = fut._loop
+
+            def complete_closure():
+                try:
+                    fut.set_result(obj)
+                except asyncio.InvalidStateError:
+                    # Avoid issues where process_notifications
+                    # and check_ready both get executed
+                    logger.debug("Failed to set result for future {}."
+                                 "It's ok.".format(fut))
+
+            loop.call_soon_threadsafe(complete_closure)
+
+        self._waiting_dict.pop(ray_object_id)
+
+    def check_immediately(self, object_id, future):
+        ready, _ = ray.wait([object_id], timeout=0)
+        if ready:
+            self._complete_future([future], object_id)
```

Comment:
I guess my concern about dictionaries was if check_immediately got called and also process_notificiations got called